ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl.hs

```diff
@@ -0,0 +1,302 @@
+        , volGetHeader        = cdbGetHeader
+        }
+    , LgrDB.LgrDbArgs {
+          lgrNodeConfig       = cdbNodeConfig
+        , lgrHasFS            = cdbHasFSLgrDB
+        , lgrDecodeLedger     = cdbDecodeLedger
+        , lgrDecodeChainState = cdbDecodeChainState
+        , lgrDecodeHash       = cdbDecodeHash
+        , lgrMemPolicy        = cdbMemPolicy
+        , lgrGenesis          = cdbGenesis
+        }
+    )
+
+-- | Internal: construct chain DB args from imm DB and vol DB
+--
+-- Useful in 'defaultArgs'
+toChainDbArgs :: ( ImmDB.ImmDbArgs m blk
+                 , VolDB.VolDbArgs m blk hdr
+                 , LgrDB.LgrDbArgs m blk
+                 )
+              -> ChainDbArgs m blk hdr
+toChainDbArgs ( ImmDB.ImmDbArgs{..}
+              , VolDB.VolDbArgs{..}
+              , LgrDB.LgrDbArgs{..}
+              ) = ChainDbArgs{
+      -- Encoders and decoders
+      cdbDecodeHash       = immDecodeHash
+    , cdbDecodeBlock      = immDecodeBlock
+    , cdbDecodeLedger     = lgrDecodeLedger
+    , cdbDecodeChainState = lgrDecodeChainState
+    , cdbEncodeHash       = immEncodeHash
+      -- Error handling
+    , cdbErrImmDb         = immErr
+    , cdbErrVolDb         = volErr
+      -- HasFS instances
+    , cdbHasFSImmDb       = immHasFS
+    , cdbHasFSVolDb       = volHasFS
+    , cdbHasFSLgrDB       = lgrHasFS
+      -- Policy
+    , cdbValidation       = immValidation
+    , cdbBlocksPerFile    = volBlocksPerFile
+    , cdbMemPolicy        = lgrMemPolicy
+      -- Integration
+    , cdbNodeConfig       = lgrNodeConfig
+    , cdbEpochSize        = immEpochSize
+    , cdbIsEBB            = immIsEBB
+    , cdbGetHeader        = volGetHeader
+    , cdbGenesis          = lgrGenesis
+    }
+
+{-------------------------------------------------------------------------------
+  Internal environment
+-------------------------------------------------------------------------------}
+
+data ChainDbEnv m blk hdr = CDB {
+      cdbImmDB   :: ImmDB m blk
+    , cdbVolDB   :: VolDB m blk hdr
+    , cdbLgrDB   :: LgrDB m blk
+    , cdbChain   :: TVar m (AnchoredFragment hdr)
+    , cdbInvalid :: TVar m (Set (Point blk))
+    , cdbHeader  :: blk -> hdr
+    }
+
+{-------------------------------------------------------------------------------
+  Initialization
+-------------------------------------------------------------------------------}
+
+openDB :: forall m blk hdr.
+          ( MonadSTM   m
+          , MonadST    m
+          , MonadCatch m
+          , HasHeader hdr
+          , HeaderHash hdr ~ HeaderHash blk
+          , ProtocolLedgerView blk
+          )
+       => ChainDbArgs m blk hdr -> m (ChainDB m blk hdr)
+openDB args = do
+    immDB   <- ImmDB.openDB argsImmDb
+    volDB   <- VolDB.openDB argsVolDb
+    lgrDB   <- LgrDB.openDB argsLgrDb
+                            immDB
+                            (implGetKnownBlock immDB volDB)
+    chain   <- atomically $ newTVar undefined
```

Comment:
We now need to do the intiial chain selection step (see ChainDB.md).

Reply:
Primary responsibilities of this layer: (1) Chain selection (on init and on `addBlock`), (2) recovery after an exception gets thrown and (3) iteration (especially `Reader`).